meta/autoid/autoid_service.go

```diff
@@ -0,0 +1,234 @@
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/autoid"
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/util/logutil"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+var _ Allocator = &singlePointAlloc{}
+
+type singlePointAlloc struct {
+	dbID          int64
+	tblID         int64
+	lastAllocated int64
+	isUnsigned    bool
+	clientDiscover
+}
+
+type clientDiscover struct {
+	// This the etcd client for service discover
+	etcdCli *clientv3.Client
+	// This is the real client for the AutoIDAlloc service
+	mu struct {
+		sync.RWMutex
+		autoid.AutoIDAllocClient
+		// Release the client conn to avoid resource leak!
+		// See https://github.com/grpc/grpc-go/issues/5321
+		*grpc.ClientConn
+	}
+}
+
+const (
+	autoIDLeaderPath = "tidb/autoid/leader"
+)
+
+func (d *clientDiscover) GetClient(ctx context.Context) (autoid.AutoIDAllocClient, error) {
+	d.mu.RLock()
+	cli := d.mu.AutoIDAllocClient
+	if cli != nil {
+		d.mu.RUnlock()
+		return cli, nil
+	}
+	d.mu.RUnlock()
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.mu.AutoIDAllocClient != nil {
+		return d.mu.AutoIDAllocClient, nil
+	}
+
+	resp, err := d.etcdCli.Get(ctx, autoIDLeaderPath, clientv3.WithFirstCreate()...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return nil, errors.New("autoid service leader not found")
+	}
+
+	addr := string(resp.Kvs[0].Value)
+	grpcConn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cli = autoid.NewAutoIDAllocClient(grpcConn)
+	d.mu.AutoIDAllocClient = cli
+	d.mu.ClientConn = grpcConn
+	return cli, nil
+}
+
+// Alloc allocs N consecutive autoID for table with tableID, returning (min, max] of the allocated autoID batch.
+// The consecutive feature is used to insert multiple rows in a statement.
+// increment & offset is used to validate the start position (the allocator's base is not always the last allocated id).
+// The returned range is (min, max]:
+// case increment=1 & offset=1: you can derive the ids like min+1, min+2... max.
+// case increment=x & offset=y: you firstly need to seek to firstID by `SeekToFirstAutoIDXXX`, then derive the IDs like firstID, firstID + increment * 2... in the caller.
+func (sp *singlePointAlloc) Alloc(ctx context.Context, n uint64, increment, offset int64) (min int64, max int64, _ error) {
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("autoid.Alloc", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+		ctx = opentracing.ContextWithSpan(ctx, span1)
+	}
+
+	if !validIncrementAndOffset(increment, offset) {
+		return 0, 0, errInvalidIncrementAndOffset.GenWithStackByArgs(increment, offset)
+	}
+
+retry:
+	cli, err := sp.GetClient(ctx)
```

Comment:
Why not save the `cli` in sp, reset it only when RPC meets error?

Reply:
Exactly as you said, see

https://github.com/pingcap/tidb/pull/38809/files#diff-f0f174acbfae2c5788672a929f39e218344beb43f732431b4fd542ce78b6a33bR63-R66